### PR TITLE
Improve serializer / deserializer for RPC

### PR DESCRIPTION
**Pull request by Geod24**
I figured this can be of interest to @TrustHenry as it touches [de]serialization quite a bit.

### CODE REVIEW

File: source/agora/common/Deserializer.d [hunk 1]
```diff
@@ -149,6 +150,17 @@ public T deserializeFull (T) (scope ubyte[] data) @safe
     return value;
 }
 
+/// Ditto
+public void deserializePart (T) (ref T record, scope DeserializeDg dg) @safe
+    if (isInputRange!T && hasLength!T)
```
Comment (AndrejMitrovic) [inline source/agora/common/Deserializer.d:155]: Does `hasLength` check if `.length` is writable though?

Comment (AndrejMitrovic) [reply source/agora/common/Deserializer.d:155]: I guess it doesn't matter much, we get a compiler error either way.

Comment (Geod24) [reply source/agora/common/Deserializer.d:155]: Don't think so

File: source/agora/common/crypto/Key.d [hunk 2]
```diff
@@ -196,6 +196,12 @@ public struct PublicKey
         dg(this.data[]);
     }
 
+    /// Support for deserialization
+    public void deserialize (scope DeserializeDg dg) @safe
+    {
+        this.data.opIndex[] = dg(this.data.opIndex.length)[];
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d:202]: What does `opIndex[]` do? This is a new idiom to me..

Comment (Geod24) [reply source/agora/common/crypto/Key.d:202]: Oh. That's actually a bug in BitBlob's operator overload I need to fix.

**Comment by AndrejMitrovic** [comment]
This LGTM. @TrustHenry what are your thoughts?

File: source/agora/common/Deserializer.d [hunk 3]
```diff
@@ -184,6 +196,12 @@ public void deserializePart (ref ubyte record, scope DeserializeDg dg)
     record = dg(record.sizeof)[0];
 }
 
+/// Ditto
+public void deserializePart (ref bool record, scope DeserializeDg dg) @safe
```
Comment (TrustHenry) [inline source/agora/common/Deserializer.d:200]: A bool type has been added to deserialize part, 
but it should also be implemented in serialize part.

**Comment by TrustHenry** [comment]
It's Very Good!

File: source/agora/common/Serializer.d [hunk 4]
```diff
@@ -55,6 +55,43 @@ unittest
     assert(serializeFull(f) == [255, 255, 255, 255, 3, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3]);
 }
 
+/*******************************************************************************
+
+    A template to check if a type has a custom serialization policy
+
```
Comment (AndrejMitrovic) [inline source/agora/common/Serializer.d:62]: Missing `Params:`

**Comment by Geod24** [comment]
This needs a rework.

**Pull request closed without merging**